tensorflow_quantum/core/ops/tfq_simulate_state_op.cc

```diff
@@ -91,13 +92,40 @@ class TfqSimulateStateOp : public tensorflow::OpKernel {
     output_shape.AddDim(output_dim_size);
     output_shape.AddDim(1 << max_num_qubits);
 
-    tensorflow::Tensor *output = nullptr;
+    tensorflow::Tensor* output = nullptr;
     OP_REQUIRES_OK(context, context->allocate_output(0, output_shape, &output));
     auto output_tensor = output->matrix<std::complex<float>>();
 
+    // Cross reference with standard google cloud compute instances
+    // Memory ~= 2 * num_threads * (2 * 64 * 2 ** num_qubits in circuits)
+    // e2s2 = 2 CPU, 8GB -> Can safely do 25 since Memory = 4GB
+    // e2s4 = 4 CPU, 16GB -> Can safely do 25 since Memory = 8GB
+    // ...
+    if (max_num_qubits < 26) {
+      ComputeSmall(num_qubits, max_num_qubits, fused_circuits, context,
+                   &output_tensor);
+    } else {
+      ComputeLarge(num_qubits, max_num_qubits, fused_circuits, context,
+                   &output_tensor);
+    }
+
+    programs.clear();
+    num_qubits.clear();
+    maps.clear();
+    qsim_circuits.clear();
+    fused_circuits.clear();
+  }
+
+ private:
+  void ComputeLarge(
+      const std::vector<int>& num_qubits, const int max_num_qubits,
+      const std::vector<std::vector<qsim::GateFused<QsimGate>>>& fused_circuits,
+      tensorflow::OpKernelContext* context,
+      tensorflow::TTypes<std::complex<float>, 1, long int>::Matrix*
+          output_tensor) {
     // Instantiate qsim objects.
     const auto tfq_for = tfq::QsimFor(context);
```

Comment:
To be sure I understand, is it correct to say that, given `tfq::QsimFor`, the `qsim::Simulator` will use multiple threads managed by TF to update many parts of the same wavefunction in parallel upon gate application?  In contrast to the parallelization in `ComputeSmall`, where each TF thread manages its own wavefunction, and in each thread a QSim simulator applies gates to only one site in that wavefunction at a time (since `qsim::SequentialFor` is given a thread count of 1 to work with inside the TF thread it is called from)?

Reply:
That is correct.